### PR TITLE
Split recalculate out into multiple functions

### PR DESCRIPTION
**Pull request by dwelch-r7**
This PR is a refactor of the `recalculate` function in `payload_set.rb` and should be landed in preparation for #17901

In #17901 I ended up having to duplicate a lot of the work that was already being done in `recalculate` so in the interest of DRY and making that review easier to read we've decided to split the refactoring out to a separate PR

# Verification Steps
- [ ] Rspec/Sanity Tests pass
- [ ] Manually try to `use` and `generate` a selection of single(stageless), staged and adapted payloads

### CODE REVIEW

File: lib/msf/core/payload_set.rb [hunk 1]
```diff
@@ -70,140 +70,226 @@ def recalculate
...
-        if ((stager_platform) and
-            (stage_platform) and
-            (stager_platform & stage_platform).empty?)
-          dlog("Stager #{stager_name} and stage #{stage_name} have incompatible platforms: #{stager_platform.names} - #{stage_platform.names}", 'core', LEV_2)
-          next
-        end
+        staged_payload = calculate_staged(stage_name: stage_name,
+                                          stager_name: stager_name,
+                                          stage_components: ip,
+                                          stager_components: op)
```
Comment (adfoster-r7) [inline lib/msf/core/payload_set.rb:102]:
```suggestion
        staged_payload = calculate_staged(stage_name: stage_name,
                                          stager_name: stager_name,
                                          stage_info: stage_info,
                                          stager_info: stager_info)
```

File: lib/msf/core/payload_set.rb [hunk 2]
```diff
@@ -70,140 +70,226 @@ def recalculate
...
-      add_single(p, name, op[5])
       new_keys.push name
 
-      _adapters.each_pair { |adapter_name, ep|
-        adapter_mod, _, adapter_platform, adapter_arch, adapter_inst = ep
-        next unless adapter_inst.compatible?(p.new)
-
-        ap = build_payload(handler, mod, adapter_mod)
+      _adapters.each_pair do |adapter_name, ep|
+        adapted_single = calculate_adapted_single(adapter_name: adapter_name,
```
Comment (adfoster-r7) [inline lib/msf/core/payload_set.rb:80]: Would this be clearer with adding a `_payload` suffix?

```suggestion
        adapted_single_payload = calculate_adapted_single_payload(adapter_name: adapter_name,
```

File: lib/msf/core/payload_set.rb [hunk 3]
```diff
@@ -70,140 +70,226 @@ def recalculate
...
+      dlog("Stager #{stager_name} and stage #{stage_name} are incompatible.", 'core', LEV_2)
+      return false
+    end
+    true
+  end
+
+  def calculate_staged(stage_name:, stager_name:, stage_components:, stager_components:)
+    # if the stager or stage has a dependency, check
+    # if they are compatible
+    return nil unless check_stage_compatibility(stager_components: stager_components,
```
Comment (adfoster-r7) [inline lib/msf/core/payload_set.rb:240]: Potentially:

```suggestion
    return nil unless stage_and_stager_compatible?(stager_components: stager_components,
```

File: lib/msf/core/payload_set.rb [hunk 4]
```diff
@@ -70,140 +70,226 @@ def recalculate
...
+    end
+
+    # If the stage has a convention, make sure it's compatible with
+    # the stager's
+    if ((stage_inst) and (stage_inst.compatible?(stager_inst) == false))
+      dlog("Stager #{stager_name} and stage #{stage_name} are incompatible.", 'core', LEV_2)
+      return false
+    end
+
+
```
Comment (adfoster-r7) [inline lib/msf/core/payload_set.rb:211]:
```suggestion
```

File: lib/msf/core/payload_set.rb (hunk 2 above)
Comment (dwelch-r7) [reply lib/msf/core/payload_set.rb:80]: I'm not against it, but this is consistent with how I did the rest, would you want them all to have `_payload` at the end?

**Pull request merged**